hw/top_earlgrey/data/top_earlgrey.sv.tpl

```diff
@@ -522,21 +537,42 @@ else:
       .tl_h_o (tl_${m["name"]}_h_h2d),
       .tl_h_i (tl_${m["name"]}_h_d2h),
     % endif
-    % for p_in in m["available_input_list"] + m["available_inout_list"]:
-      % if loop.first:
+    % if m["type"] == "usbdev":
+
+      // Differential data
+      .cio_d_i          (1'b0),
+      .cio_d_o          (),
+      .cio_se0_o        (),
+
+      // Single-ended data
+      .cio_dp_i         (cio_usbdev_dp_p2d),
+      .cio_dn_i         (cio_usbdev_dn_p2d),
+      .cio_dp_o         (cio_usbdev_dp_d2p),
+      .cio_dn_o         (cio_usbdev_dn_d2p),
+
+      // Non-data I/O
+      .cio_sense_i      (cio_usbdev_sense_p2d),
+      .cio_oe_o         (cio_usbdev_dp_en_d2p),
+      .cio_tx_mode_se_o (),
+      .cio_pullup_en_o  (cio_usbdev_pullup_en_d2p),
+      .cio_suspend_o    (),
```

Comment:
What would be a reason that usbdev cannot be generalized as other IPs? I want to hear your feedback and set a plan to revise topgen to support (inter-module connection feature is coming)

Reply:
I think this was added because Stefan needs to support a real USB PHY, rather than the single ended I/Os that are also supported (and the only way on FPGA).

Could the correct thing here be a usbphy module that instantiates one or the other but takes all the signals in either case? Would be inter-module connections.

Reply:
> What would be a reason that usbdev cannot be generalized as other IPs?

There were a couple of issues related to the interface of usbdev being slightly different from other IPs such as:
- Shared output enable (`cio_oe_o` controls `cio_d_o`, `cio_se0_o`, `cio_dp_o`, `cio_dn_o`)
- Missing output enables, or "data" outputs (`cio_suspend_o`, `cio_pullup_en_o`)
- Having differential module I/Os that are connected to top-level ports.

The reason for these differences were not clear to me and I did not want to change the interface or the tools in this PR yet. The goal is to get module in as it is (more or less), have CI to make sure it does not break immediately again and then starting the work on the device and tools. There is quite some work to be done (see #1403) but I think we should break it down into smaller junks.

> Could the correct thing here be a usbphy module that instantiates one or the other but takes all the signals in either case?

I am not sure about this. Would this usbphy module have combined single-ended and differential I/Os but e.g. not drive the `*n_o` in differential mode? What would the usbphy be conntected to?